In [2]:
import pandas as pd

df = pd.read_csv("energydata_complete.csv")
df

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
...                    ...         ...     ...        ...        ...   
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3         T4  ...         T9  \
0      19.200000  44.790000  19.790000  44.730000  19.000000  ...  17.033333   
1      19.200000  44.722500  19.790000  44.790000  19.000000  ...  17.066667   
2      19.200000  44.626667  19.790000  44.933333  18.926667  ...  17.000000   
3      19.200000  44.590000  19.790000  45.000000  18.890000  ...  17.000000   
4      19.200000  44.530000  19.790000  45.000000  18.890000  ...  17.000000   
...          ...        ...        ...        ...        ...  ...        ...   
19730  25.890000  42.025714  27.200000  41.163333  24.700000  ...  23.200000   
19731  25.754000  42.080000  27.133333  41.223333  24.700000  ...  23.200000   
19732  25.628571  42.768571  27.050000  41.690000  24.700000  ...  23.200000   
19733  25.414000  43.036000  26.890000  41.290000  24.700000  ...  23.200000   
19734  25.264286  42.971429  26.823333  41.156667  24.700000  ...  23.200000   

          RH_9      T_out  Press_mm_hg     RH_out  Windspeed  Visibility  \
0      45.5300   6.600000        733.5  92.000000   7.000000   63.000000   
1      45.5600   6.483333        733.6  92.000000   6.666667   59.166667   
2      45.5000   6.366667        733.7  92.000000   6.333333   55.333333   
3      45.4000   6.250000        733.8  92.000000   6.000000   51.500000   
4      45.4000   6.133333        733.9  92.000000   5.666667   47.666667   
...        ...        ...          ...        ...        ...         ...   
19730  46.7900  22.733333        755.2  55.666667   3.333333   23.666667   
19731  46.7900  22.600000        755.2  56.000000   3.500000   24.500000   
19732  46.7900  22.466667        755.2  56.333333   3.666667   25.333333   
19733  46.8175  22.333333        755.2  56.666667   3.833333   26.166667   
19734  46.8450  22.200000        755.2  57.000000   4.000000   27.000000   

       Tdewpoint        rv1        rv2  
0       5.300000  13.275433  13.275433  
1       5.200000  18.606195  18.606195  
2       5.100000  28.642668  28.642668  
3       5.000000  45.410389  45.410389  
4       4.900000  10.084097  10.084097  
...          ...        ...        ...  
19730  13.333333  43.096812  43.096812  
19731  13.300000  49.282940  49.282940  
19732  13.266667  29.199117  29.199117  
19733  13.233333   6.322784   6.322784  
19734  13.200000  34.118851  34.118851  

[19735 rows x 29 columns]

The dataset contains various columns, including:

- 'date': Timestamp representing the date and time.
- 'Appliances': The number of household appliances in use.
- 'lights': The number of lights in use.
- Several other columns such as 'T1', 'RH_1', 'T2', 'RH_2', 'T3', 'RH_3', 'T4', 'RH_4', and more, which likely represent temperature and humidity measurements for different areas or rooms.

The dataset seems to capture various environmental and energy-related measurements.

Dataset Information
https://archive.ics.uci.edu/ml/machine-learning-databases/00374

The dataset for the remainder of this quiz (from question 18) is the Appliances Energy Prediction data. The data set is at 10 min for about 4.5 months. The house temperature and humidity conditions were monitored with a ZigBee wireless sensor network. Each wireless node transmitted the temperature and humidity conditions around 3.3 min. Then, the wireless data was averaged for 10 minutes periods. The energy data was logged every 10 minutes with m-bus energy meters. Weather from the nearest airport weather station (Chievres Airport, Belgium) was downloaded from a public data set from Reliable Prognosis (rp5.ru), and merged together with the experimental data sets using the date and time column. Two random variables have been included in the data set for testing the regression models and to filter out non predictive attributes (parameters). The attribute information can be seen below.

Attribute Information:

Date, time year-month-day hour:minute:second

Appliances, energy use in Wh 

lights, energy use of light fixtures in the house in Wh

T1, Temperature in kitchen area, in Celsius

RH_1, Humidity in kitchen area, in %

T2, Temperature in living room area, in Celsius

RH_2, Humidity in living room area, in %

T3, Temperature in laundry room area

RH_3, Humidity in laundry room area, in %

T4, Temperature in office room, in Celsius

RH_4, Humidity in office room, in %

T5, Temperature in bathroom, in Celsius

RH_5, Humidity in bathroom, in %

T6, Temperature outside the building (north side), in Celsius

RH_6, Humidity outside the building (north side), in %

T7, Temperature in ironing room , in Celsius

RH_7, Humidity in ironing room, in %

T8, Temperature in teenager room 2, in Celsius

RH_8, Humidity in teenager room 2, in %

T9, Temperature in parents room, in Celsius

RH_9, Humidity in parents room, in %

To, Temperature outside (from Chievres weather station), in Celsius

Pressure (from Chievres weather station), in mm Hg

RH_out, Humidity outside (from Chievres weather station), in %

Wind speed (from Chievres weather station), in m/s

Visibility (from Chievres weather station), in km

Tdewpoint (from Chievres weather station), °C

rv1, Random variable 1, nondimensional

rv2, Random variable 2, nondimensional

Question 17
From the dataset, fit a linear model on the relationship between the temperature in
the living room in Celsius (x = T2) and the temperature outside the building (y = T6).
What is the Root Mean Squared error in three D.P?
4.644
6.644
7.654
3.644

In [3]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Select the relevant columns
X = df[['T2']].values  # Temperature in the living room (independent variable)
y = df['T6'].values    # Temperature outside the building (dependent variable)

# Create and fit the linear regression model
model = LinearRegression()
model.fit(X, y)

# Make predictions
y_pred = model.predict(X)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y, y_pred))

# Print the RMSE rounded to 3 decimal places
print(round(rmse, 3))


3.644


Question 18
Remove the following columns: ["date", "lights"]. The target variable is "Appliances".
Use a 70-30 train-test set split with a random state of 42 (for reproducibility).
Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler
fit_transform and transform methods on the train and test set respectively). Run a
multiple linear regression using the training set. Answer the following questions:
What is the Mean Absolute Error (in three decimal places) for the training set?
52.899
55.745
53.747
50.777

In [13]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a multiple linear regression model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the training set
y_train_pred = model.predict(X_train)

# Calculate the Mean Absolute Error (MAE) on the training set
mae_train = mean_absolute_error(y_train, y_train_pred)

# Print the MAE rounded to 5 decimal places
print(round(mae_train, 5))


53.74228


In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a multiple linear regression model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the training set
y_train_pred = model.predict(X_train)

# Calculate the Mean Absolute Error (MAE) on the training set
mae_train = mean_absolute_error(y_train, y_train_pred)

# Print the MAE rounded to 5 decimal places
print(round(mae_train, 5))


53.74228


Question 19
What is the Root Mean Squared Error (in three decimal places) for the training set?
94.221
95.216
89.346
99.454

In [6]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import math

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Train a multiple linear regression model on the training set
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Make predictions on the training set
y_train_pred = model.predict(X_train_scaled)

# Calculate the Mean Squared Error (MSE) on the training set
mse_train = mean_squared_error(y_train, y_train_pred)

# Calculate the RMSE (Root Mean Squared Error) on the training set
rmse_train = math.sqrt(mse_train)

# Print the RMSE rounded to 3 decimal places
print(round(rmse_train, 3))



95.216


Question 20
What is the Mean Absolute Error (in three decimal places) for test set?
55.657
56.667
60.767
53.647

In [7]:
from sklearn.metrics import mean_absolute_error

# Normalize the test set using the same MinMaxScaler used for the training set
X_test_scaled = scaler.transform(X_test)

# Make predictions on the test set
y_test_pred = model.predict(X_test_scaled)

# Calculate the Mean Absolute Error (MAE) on the test set
mae_test = mean_absolute_error(y_test, y_test_pred)

# Print the MAE rounded to 3 decimal places
print(round(mae_test, 3))


53.643


Question 21
What is the Root Mean Squared Error (in three decimal places) for test set?
90.64
93.64
91.64
99.76

In [8]:
from sklearn.metrics import mean_squared_error
import math

# Normalize the test set using the same MinMaxScaler used for the training set
X_test_scaled = scaler.transform(X_test)

# Make predictions on the test set
y_test_pred = model.predict(X_test_scaled)

# Calculate the Mean Squared Error (MSE) on the test set
mse_test = mean_squared_error(y_test, y_test_pred)

# Calculate the RMSE by taking the square root of the MSE
rmse_test = math.sqrt(mse_test)

# Print the RMSE rounded to 3 decimal places
print(round(rmse_test, 3))


93.64


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a multiple linear regression model on the training set
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Make predictions on both training and test sets
y_train_pred = model.predict(X_train_scaled)
y_test_pred = model.predict(X_test_scaled)

# Calculate the Mean Absolute Error (MAE) on both sets
mae_train = mean_absolute_error(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

# Print the MAE for both training and test sets
print(f"Training MAE: {mae_train:.3f}")
print(f"Test MAE: {mae_test:.3f}")


Question 23
Train a ridge regression model with default parameters. Is there any change to the
root mean squared error (RMSE) when evaluated on the test set?
Yes
No

In [9]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error
import math

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a Ridge regression model on the training set with default parameters
ridge_model = Ridge()
ridge_model.fit(X_train_scaled, y_train)

# Make predictions on the test set
y_test_pred_ridge = ridge_model.predict(X_test_scaled)

# Calculate the Mean Squared Error (MSE) on the test set
mse_test_ridge = mean_squared_error(y_test, y_test_pred_ridge)

# Calculate the RMSE by taking the square root of the MSE
rmse_test_ridge = math.sqrt(mse_test_ridge)

# Load the previous RMSE from Question 21
previous_rmse = 93.64

# Compare the RMSE of the Ridge regression model with the previous RMSE
if round(rmse_test_ridge, 3) == round(previous_rmse, 3):
    print("No change in RMSE.")
else:
    print("There is a change in RMSE.")


There is a change in RMSE.


Question 24
Train a lasso regression model with default value and obtain the new feature weights
with it. How many of the features have non-zero feature weights?
6 or 7
20 or 21
4 or 5
25 or 26

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Lasso

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a Lasso regression model on the training set with default parameters
lasso_model = Lasso()
lasso_model.fit(X_train_scaled, y_train)

# Get the coefficients (weights) of the features
feature_weights = lasso_model.coef_

# Count the number of features with non-zero weights
non_zero_features = sum(feature_weights != 0)

print("Number of features with non-zero feature weights:", non_zero_features)


Number of features with non-zero feature weights: 4


Question 25
What is the new RMSE with the Lasso Regression on the test set?
93.892
91.892
89.932
99.992

In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error
import math

# Load the dataset
df = pd.read_csv("energydata_complete.csv")

# Remove specified columns
df = df.drop(columns=["date", "lights"])

# Define features (X) and target variable (y)
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

# Split the dataset into a training set and a test set (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a Lasso regression model on the training set with default parameters
lasso_model = Lasso()
lasso_model.fit(X_train_scaled, y_train)

# Make predictions on the test set
y_test_pred = lasso_model.predict(X_test_scaled)

# Calculate the Mean Squared Error (MSE) on the test set
mse_test = mean_squared_error(y_test, y_test_pred)

# Calculate the RMSE by taking the square root of the MSE
rmse_test = math.sqrt(mse_test)

# Print the new RMSE rounded to 3 decimal places
print(round(rmse_test, 3))


99.424
